### ISSUE TITLE
Additional eigenproblems decompositions

### ISSUE THREAD

**Issue by 0x00b1** [pytorch/pytorch#78809]
### 🚀 The feature, motivation and pitch

There are a handful of eigenproblem decompositions that I would like to use but are not provided by PyTorch (e.g. Hessenberg, Jordan, and Schur). This is one area where PyTorch lags behind MATLAB.

**Comment by nikitaved** [comment]
I think we can have these for the CPU as Lapack has it. For CUDA we would need to implement them from scratch. At least for sure the Schure decomposition is not there in MAGMA, which blocked me on implementing the matrix square root in a non-iterative way...

**Comment by Lezcano** [comment]
@xwang233 for new algorithms that cuSOLVER could implement.

**Comment by Lezcano** [comment]
fwiw, note that Jordan is like `linalg.eig` for almost all matrices. They only differ for non-diagonalisable matrices, and for that to happen at least one eigenvalue must be repeated so, while it's an interesting decomposition on its own, its less so in the context of floating point precision.

**Comment by 0x00b1** [comment]
> fwiw, note that Jordan is like linalg.eig for almost all matrices. They only differ for non-diagonalisable matrices, and for that to happen at least one eigenvalue must be repeated so, while it's an interesting decomposition on its own, its less so in the context of floating point precision.

Yeah. I should have omitted Jordan. I was trying to enumerate some obvious examples. In my experience, the Schur decomposition is pretty much always preferable!

**Comment by Lezcano** [comment]
Yup. Also, with the schur decomposition we could implement some inverse matrix functions, like the sqrt or the matrix logarithm, which is quite cool

**Comment by nikitaved** [comment]
Related: https://github.com/pytorch/pytorch/issues/9983. Having them implemented through the Schur decomposition is probably our best bet in terms of accuracy, then performance.

**Comment by xwang233** [comment]
Thank you for sending this feature request. I just forwarded the request to cusolver team.

Related: https://github.com/pytorch/pytorch/issues/78581

**Comment by opooladz** [comment]
A Hessenberg-Schur method for the problem AX + XB= C via [this](https://ieeexplore.ieee.org/document/1102170) would be nice.

**Comment by lezcano** [comment]
magma and cusolver don't have an implementation, so you should probably first ask for it there

**Comment by hazdzz** [comment]
Does the PyTorch team have plans to implement these two types of matrix decomposition?

**Comment by rgjha** [comment]
It would be very nice if PyTorch can implement these. Would be great for physicists in general :)

**Comment by lezcano** [comment]
I think https://github.com/pytorch/pytorch/issues/78809#issuecomment-1850868304 is still relevant
